cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/CaReconciler.java

```diff
@@ -374,39 +374,35 @@ Future<Void> reconcileClusterOperatorSecret(Clock clock) {
+     * It is not necessary when the CA certificate is renewed while retaining the existing key.
+     *
+     * If Strimzi did not replace the CA key during the current reconciliation, {@code isClusterCaNeedFullTrust} is used to:
+     *      * continue from a previous CA key replacement which didn't end successfully (i.e. CO stopped)
+     *      * track key replacements when the user is managing the CA
+     *
+     * @return Future which completes when this step is done, either by rolling the cluster or by deciding
+     *         that no rolling is needed.
      */
-    Future<Void> rollingUpdateForNewCaKey() {
-        RestartReasons podRollReasons = RestartReasons.empty();
-
-        // cluster CA needs to be fully trusted
-        // it is coming from a cluster CA key replacement which didn't end successfully (i.e. CO stopped) and we need to continue from there
+    Future<Void> maybeRollingUpdateForNewClusterCaKey() {
         if (clusterCa.keyReplaced() || isClusterCaNeedFullTrust) {
-            podRollReasons.add(RestartReason.CLUSTER_CA_CERT_KEY_REPLACED);
-        }
-
-        if (clientsCa.keyReplaced()) {
-            podRollReasons.add(RestartReason.CLIENT_CA_CERT_KEY_REPLACED);
-        }
-
-        if (podRollReasons.shouldRestart()) {
+            String restartReason = RestartReason.CLUSTER_CA_CERT_KEY_REPLACED.getDefaultNote();
             TlsPemIdentity coTlsPemIdentity = new TlsPemIdentity(new PemTrustSet(clusterCa.caCertSecret()), PemAuthIdentity.clusterOperator(coSecret));
             return getZooKeeperReplicas()
-                    .compose(replicas -> maybeRollZookeeper(replicas, podRollReasons, coTlsPemIdentity))
+                    .compose(replicas -> rollZookeeper(replicas, restartReason, coTlsPemIdentity))
                     .compose(i -> getKafkaReplicas())
-                    .compose(nodes -> rollKafkaBrokers(nodes, podRollReasons, coTlsPemIdentity))
-                    .compose(i -> maybeRollDeploymentIfExists(KafkaResources.entityOperatorDeploymentName(reconciliation.name()), podRollReasons))
-                    .compose(i -> maybeRollDeploymentIfExists(KafkaExporterResources.componentName(reconciliation.name()), podRollReasons))
-                    .compose(i -> maybeRollDeploymentIfExists(CruiseControlResources.componentName(reconciliation.name()), podRollReasons));
+                    .compose(nodes -> rollKafkaBrokers(nodes, RestartReasons.of(RestartReason.CLUSTER_CA_CERT_KEY_REPLACED), coTlsPemIdentity))
+                    .compose(i -> rollDeploymentIfExists(KafkaResources.entityOperatorDeploymentName(reconciliation.name()), restartReason))
+                    .compose(i -> rollDeploymentIfExists(KafkaExporterResources.componentName(reconciliation.name()), restartReason))
+                    .compose(i -> rollDeploymentIfExists(CruiseControlResources.componentName(reconciliation.name()), restartReason));
```

Comment:
Can we stick to use the restartReason the same way everywhere? Not once as a String, once as a RestartReason etc.? Or remove it as a parameter from the methods where it is the only option used anyway? The way it is done now seems very weird... you prepare the String, but then you don't use it everywhere etc.